language_service/src/hover.rs

```diff
@@ -188,18 +193,38 @@ impl Visitor<'_> for HoverVisitor<'_> {
 }
 
 fn markdown_with_doc(doc: &Rc<str>, code: impl Display) -> String {
-    if doc.is_empty() {
+    let parsed_doc = parse_doc(doc);
+    if parsed_doc.summary.is_empty() {
         markdown_fenced_block(code)
     } else {
         format!(
             "{}
 {}",
-            doc,
+            parsed_doc.summary,
             markdown_fenced_block(code)
         )
     }
 }
 
+fn parse_doc(doc: &str) -> Documentation {
+    let summary_re =
+        Regex::new(r"(^|(\r?\n))\s*#\s*((S|s)ummary)[\s\r\n]*").expect("Invalid regex");
```

Comment:
These seem a little permissive. I believe there shouldn't be any whitespace after the newline and before `# Summary`, and probably shouldn't be any trailing white-space, so you can lose some of the `\s*`s. Also, you probably want to require a newline at the end (the `*` makes it optional, so it would match `   # Summary-like`).

Using the flags support, enabling multi-line will also clean-up the need for the `\r\n` characters (just use ^ and $ to match the start/end of lines), and just make it case-insensitive to ignore Summary vs summary vs SUMMARY.

Ultimately, something like `(?mi)^# ?summary$`.

Reply:
(And actually lose the ? in there. There should always be a space after the '#' I believe. i.e. `# Header` is valid, `#Header` shouldn't be).

Reply:
Yah the use of the flags to enable multi-line really cleans up the regex a lot. I've used a variation of the regex you proposed below.